### PR TITLE
Cert hash data refactoring

### PR DESCRIPTION
**Pull request by abi87**

### CODE REVIEW

File: src/coins.h [hunk 1]
```diff
@@ -701,18 +698,19 @@ class CCoinsViewCache : public CCoinsViewBacked
...
-    bool GetCertDataHash(const uint256& scId,
-                         const int epoch,
-                         libzendoomc::ScFieldElement& certDataHash) const override;
-    bool GetCertDataCumulativeHash(const uint256& scId,
-                                   const int epoch,
-                                   libzendoomc::ScFieldElement& certDataCumulativeHash) const override;
+//    bool GetCertDataHash(const uint256& scId, const int epoch,
+//                         libzendoomc::ScFieldElement& certDataHash) const ;
+//    bool GetCertDataCumulativeHash(const uint256& scId, const int epoch,
+//                                   libzendoomc::ScFieldElement& prevEpochCumulativeCertDataHash) const;
```
Comment (VolodymyrBabaiev) [inline src/coins.h:713]: I think we can remove it now.

File: src/coins.cpp [hunk 2]
```diff
@@ -379,6 +382,24 @@ CSidechainEventsMap::iterator CCoinsViewCache::ModifySidechainEvents(int height)
...
+
+    std::pair<libzendoomc::ScFieldElement, libzendoomc::ScFieldElement> baseData;
+    if (!base->GetCertDataHashes(scId, epoch, baseData))
+        return cacheCertDataHashes.end();
+
+    //Fill cache and return iterator. The insert in cache below looks cumbersome. However
+    //it allows to insert CCeasingSidechains and keep iterator to inserted member without extra searches
+    CCertDataHashMap::iterator ret = cacheCertDataHashes.insert(std::make_pair(key, CCertDataHashCacheEntry(baseData, CCertDataHashCacheEntry::Flags::DEFAULT ))).first;
+
+    //cachedCoinsUsage += ret->second.scEvents.DynamicMemoryUsage(); //TODO
```
Comment (VolodymyrBabaiev) [inline src/coins.cpp:399]: Do we need this one here?

File: src/coins.cpp [hunk 3]
```diff
@@ -573,96 +594,49 @@ bool CCoinsViewCache::HaveCswNullifier(const uint256& scId,
...
-
-    std::pair<uint256, int> position = std::make_pair(scId, epoch);
-
-    CCertDataHashMap::iterator it = cacheCertDataHashes.find(position);
-    if (it != cacheCertDataHashes.end()) {
-        certDataCumulativeHash = it->second.certDataCumulativeHash;
+bool CCoinsViewCache::GetCertDataHashes(const uint256& scId, const int epoch,
+                       std::pair<libzendoomc::ScFieldElement, libzendoomc::ScFieldElement>& certDataHashes) const
+{
+    CCertDataHashMap::const_iterator it = FetchCertDataEntry(scId, epoch);
```
Comment (VolodymyrBabaiev) [inline src/coins.cpp:632]: We also have to use this function in RemoveCertDataHash.

File: src/coins.h (hunk 1 above)
Comment (abi87) [reply src/coins.h]: Done

File: src/coins.cpp (hunk 3 above)
Comment (abi87) [reply src/coins.cpp:609]: Agree, done

File: src/coins.cpp (hunk 2 above)
Comment (abi87) [reply src/coins.cpp:400]: We need to account for the DynamicMemoryUsage of CertDataHashes. Going to fix this in a different PR

File: src/coins.cpp [hunk 4]
```diff
@@ -1203,14 +1217,20 @@ bool CCoinsViewCache::IsTxCswApplicableToState(const CTransaction& tx, CValidati
...
-        const libzendoomc::ScFieldElement& currentCertDataHash = libzendoomc::ScFieldElement(); // GetCertDataHash(csw.scId, csw.nEpoch);
-        const libzendoomc::ScFieldElement& lastCumulativeCertDataHash = libzendoomc::ScFieldElement(); // GetLastCumulativeCertDataHash(csw.scId);
+        std::pair<libzendoomc::ScFieldElement, libzendoomc::ScFieldElement> certDataHashes;
+        if (!this->GetCertDataHashes(csw.scId, csw.nEpoch, certDataHashes))
+            return state.Invalid(error("missing cert data hash for required scId/epoch"),
+                 REJECT_INVALID, "tx-csw-missing-cert-data-hash");
+
+        const libzendoomc::ScFieldElement& prevCumulativeCertDataHash = certDataHashes.second;
+        const libzendoomc::ScFieldElement& currentCertDataHash = certDataHashes.first;
+        libzendoomc::ScFieldElement lastCumulativeCertDataHash = libzendoomc::ScFieldElement();
```
Comment (i-Alex) [inline src/coins.cpp:1227]: We need to get hashes for the most recent epoch valid Certificate -> then calculate the cumulative one.
For example, SC ceased at epoch 100, so the last Certificate was submitted in epoch 99, but will be invalidated.
So, the most recent one was submitted in epoch 98 for epoch 97.
We need to call `GetCertDataHashes(csw.scId, 97, recentDataHashes)`
I guess that `97` is taken from CSidechain entry for given `scId`.

File: src/coins.h [hunk 5]
```diff
@@ -750,24 +741,19 @@ class CCoinsViewCache : public CCoinsViewBacked
...
-    CSidechainEventsMap::const_iterator FetchSidechainEvents(int height) const;
-    CSidechainEventsMap::iterator  ModifySidechainEvents(int height);
+    CCoinsMap::const_iterator           FetchCoins(const uint256 &txid)       const;
+    CCoinsMap::iterator                 FetchCoins(const uint256 &txid);
+    CSidechainsMap::const_iterator      FetchSidechains(const uint256& scId)  const;
+    CSidechainsMap::iterator            ModifySidechain(const uint256& scId);
+    const CSidechain* const             AccessSidechain(const uint256& scId);
+    CSidechainEventsMap::const_iterator FetchSidechainEvents(int height)      const;
+    CSidechainEventsMap::iterator       ModifySidechainEvents(int height);
+    CCertDataHashMap::const_iterator    FetchCertDataEntry(const uint256& scId, const int epoch) const;
```
Comment (i-Alex) [inline src/coins.h:751]: Add `ModifyCertDataEntry` to make CerDataEntries modification similar to other approaches (ModifySidechainEvents, ModifySidechain).

**Review by i-Alex — CHANGES_REQUESTED** [review]

File: src/coins.cpp [hunk 6]
```diff
@@ -559,7 +581,7 @@ bool CCoinsViewCache::HaveCswNullifier(const uint256& scId,
     
     CCswNullifiersMap::iterator it = cacheCswNullifiers.find(position);
     if (it != cacheCswNullifiers.end())
-        return (it->second.flag = CCswNullifiersCacheEntry::ERASED);
+        return (it->second.flag == CCswNullifiersCacheEntry::ERASED);
```
[the remark this answers is not in the document]
Comment (abi87) [reply src/coins.cpp:584]: Going to fix in csw_nullifier_refactoring

File: src/coins.cpp [hunk 7]
```diff
@@ -572,105 +594,92 @@ bool CCoinsViewCache::HaveCswNullifier(const uint256& scId,
...
-        ret.first->second.certDataCumulativeHash = tmp;
-    }
+bool CCoinsViewCache::UpdateCertDataHash(const CTxScCreationOut& scCreationOut) {
+    if(!HaveSidechain(scCreationOut.GetScId()))
+        return false;
 
-    return res;
+    libzendoomc::ScFieldElement nullCertDataHash{};
+    CCertDataHashCacheEntry newEntry{std::make_pair(nullCertDataHash, nullCertDataHash),CCertDataHashCacheEntry::Flags::FRESH};
+    this->cacheCertDataHashes[std::make_pair(scCreationOut.GetScId(), 0)] = newEntry;
```
[the remark this answers is not in the document]
Comment (abi87) [reply src/coins.cpp]: Agree, done

File: src/coins.cpp [hunk 8]
```diff
@@ -572,105 +594,92 @@ bool CCoinsViewCache::HaveCswNullifier(const uint256& scId,
...
-        ret.first->second.certDataCumulativeHash = tmp;
-        base->GetCertDataHash(scId, epoch, tmp);
-        ret.first->second.certDataHash = tmp;
+    if (HaveCertDataHashes(cert.GetScId(), cert.epochNumber))
+    {
+        //pick or create scId entry in scUndoDatabyScId
+        blockUndo.scUndoDatabyScId[cert.GetScId()].prevTopCommittedCertDataHash = this->cacheCertDataHashes.at(std::make_pair(cert.GetScId(), cert.epochNumber)).certDataHash;
+        blockUndo.scUndoDatabyScId[cert.GetScId()].contentBitMask |= CSidechainUndoData::AvailableSections::CERT_DATA_HASH;
+
+        this->cacheCertDataHashes.at(std::make_pair(cert.GetScId(), cert.epochNumber)).certDataHash = certDataHash;
```
[the remark this answers is not in the document]
Comment (abi87) [reply src/coins.cpp]: Agree, done

File: src/coins.cpp [hunk 9]
```diff
@@ -572,105 +594,92 @@ bool CCoinsViewCache::HaveCswNullifier(const uint256& scId,
...
-        
-        std::pair<CCertDataHashMap::iterator, bool> ret = cacheCertDataHashes.insert(std::make_pair(position, CCertDataHashCacheEntry()));
-        ret.first->second.flag = CCertDataHashCacheEntry::DEFAULT;
-        ret.first->second.certDataCumulativeHash = tmp;
-        base->GetCertDataHash(scId, epoch, tmp);
-        ret.first->second.certDataHash = tmp;
+    if (HaveCertDataHashes(cert.GetScId(), cert.epochNumber))
+    {
+        //pick or create scId entry in scUndoDatabyScId
+        blockUndo.scUndoDatabyScId[cert.GetScId()].prevTopCommittedCertDataHash = this->cacheCertDataHashes.at(std::make_pair(cert.GetScId(), cert.epochNumber)).certDataHash;
```
[the remark this answers is not in the document]
Comment (abi87) [reply src/coins.cpp]: Done

File: src/coins.cpp (hunk 4 above)
Comment (abi87) [reply src/coins.cpp:1241]: Done

File: src/coins.h (hunk 5 above)
Comment (abi87) [reply src/coins.h:749]: Done

**Review by i-Alex — APPROVED** [review]
Very nice refactoring, especially UTs!

**Review by alsala — APPROVED** [review]

**Pull request merged**